README.rst
.. image:: https://img.shields.io/badge/License-GPL%20v3-blue.svg
   :target: https://www.gnu.org/licenses/gpl-3.0

.. image:: https://badge.fury.io/py/storage-cache.svg
   :target: https://pypi.org/project/storage-cache/

.. image:: https://img.shields.io/travis/HenleyKuang/storage-cache.svg
   :target: https://travis-ci.org/HenleyKuang/storage-cache


=============
Storage Cache
=============

Installation
------------

.. code-block:: bash

    pip install storage-cache


How to use
----------

.. code-block:: python

    from storage_cache import StorageCacheFactory
    from storage_cache.types import REDIS_STORAGE_CACHE

    StorageCache = StorageCacheFactory.factory(REDIS_STORAGE_CACHE)
    storage_cache = StorageCache(
        <redis_host>, <redis_port>, <redis_password>, <redis_db>, <key_format>)


CHANGES.rst
CHANGES ( Latest to Oldest )
============================

v1.1.0
------

- Change name to storage-cache

v1.0.0
------

Inition version.



tests/test_redis_cache.py
"""
:author: Henley Kuang
:since: 2019-06-14
"""
import json
import pytest

from storage_cache.main import StorageCacheFactory
from storage_cache.types import REDIS_STORAGE_CACHE
from storage_cache.excepts import CacheClassDoesNotExistException

PRIMARY_KEYS = ['primary', 'keys', 'test']
SECONDARY_KEYS = ['secondary', 'keys', 'hashfield']
META_DATA = ['meta', 'data', 'value']
META_DATA_STR = json.dumps(META_DATA)

EXPIRATION = 3600 * 24 * 7 * 2  # 2 weeks

HASH_KEY_PREFIX = 'hash:unittest_prefix'
STR_KEY_PREFIX = 'str:unittest_prefix'

StorageCacheClass = StorageCacheFactory.factory(REDIS_STORAGE_CACHE)
storage_cache_hash = StorageCacheClass(redis_host='127.0.0.1',
                                       redis_port=6379,
                                       redis_password='',
                                       redis_db=0,
                                       key_format=HASH_KEY_PREFIX,
                                       )
storage_cache_str = StorageCacheClass(redis_host='127.0.0.1',
                                      redis_port=6379,
                                      redis_password='',
                                      redis_db=0,
                                      key_format=STR_KEY_PREFIX,
                                      )


def test_set_hash_cache():
    """
    Test return value of setting hash data type
    """
    result = storage_cache_hash.set_cache_hash(
        PRIMARY_KEYS, SECONDARY_KEYS, META_DATA_STR, EXPIRATION)
    result_cmd, result_hash_key_name, result_hash_field_name, result_meta_data_str = result
    result_meta_data = json.loads(result_meta_data_str)
    expected_cmd = 'hset'
    expected_hash_key_name = "%s:%s" % (
        HASH_KEY_PREFIX, ":".join(PRIMARY_KEYS))
    expected_hash_field_name = ":".join(SECONDARY_KEYS)
    expected_meta_data = META_DATA
    assert isinstance(result, tuple)
    assert result_cmd == expected_cmd
    assert result_hash_key_name == expected_hash_key_name
    assert result_hash_field_name == expected_hash_field_name
    assert result_meta_data == expected_meta_data


def test_set_str_cache():
    """
    Test return value of setting str date type
    """
    result = storage_cache_str.set_cache_str(
        PRIMARY_KEYS, META_DATA_STR, EXPIRATION)
    result_cmd, result_str_key_name, result_meta_data_str = result
    result_meta_data = json.loads(result_meta_data_str)
    expected_cmd = 'set'
    expected_str_key_name = "%s:%s" % (STR_KEY_PREFIX, ":".join(PRIMARY_KEYS))
    expected_meta_data = META_DATA
    assert isinstance(result, tuple)
    assert result_cmd == expected_cmd
    assert result_str_key_name == expected_str_key_name
    assert result_meta_data == expected_meta_data


def test_get_hash_cache():
    """
    Test return value of getting hash date type
    """
    result = storage_cache_hash.get_cache_hash(
        PRIMARY_KEYS, SECONDARY_KEYS)
    result_meta_data_str = result
    result_meta_data = json.loads(result_meta_data_str)
    expected_meta_data = META_DATA
    assert isinstance(result, str)
    assert result_meta_data == expected_meta_data


def test_get_str_cache():
    """
    Test return value of getting str data type
    """
    result = storage_cache_str.get_cache_str(
        PRIMARY_KEYS)
    result_meta_data_str = result
    result_meta_data = json.loads(result_meta_data_str)
    expected_meta_data = META_DATA
    assert isinstance(result, str)
    assert result_meta_data == expected_meta_data


playground/storage_cache_playground.py
import argparse
import logging
import json

from storage_cache import StorageCacheFactory
from storage_cache.util.config import DEFAULT_LOG_FORMAT_STRING

LOGGER = logging.getLogger(__name__)


def _parse_args():
    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawTextHelpFormatter)

    base_parser = argparse.ArgumentParser(add_help=False)
    base_parser.add_argument('--cache-type',  required=True)
    base_parser.add_argument('--redis-host', required=True)
    base_parser.add_argument('--redis-port', required=True)
    base_parser.add_argument('--redis-password')
    base_parser.add_argument('--redis-db', required=True)
    base_parser.add_argument('--key-format', required=True, help='key prefix')
    base_parser.add_argument('--primary-keys', nargs='+',
                             required=True, help='list of strings')

    hash_parser = argparse.ArgumentParser(add_help=False)
    hash_parser.add_argument('--secondary-keys', nargs='+',
                             required=True, help='list of strings')

    set_parser = argparse.ArgumentParser(add_help=False)
    set_parser.add_argument('--meta-data', nargs='+',
                            required=True, help='list of strings')

    command_subparser = parser.add_subparsers(dest='sub_command')
    command_subparser.required = True

    command_subparser.add_parser(
        'set-hash', parents=[base_parser, hash_parser, set_parser])
    command_subparser.add_parser(
        'get-hash', parents=[base_parser, hash_parser])
    command_subparser.add_parser('set-str', parents=[base_parser, set_parser])
    command_subparser.add_parser('get-str', parents=[base_parser])

    return parser.parse_args()


def _main():
    options = _parse_args()
    cache_type = options.cache_type
    redis_host = options.redis_host
    redis_port = options.redis_port
    redis_password = options.redis_password
    redis_db = options.redis_db
    key_format = options.key_format
    primary_keys = options.primary_keys

    StorageCache = StorageCacheFactory.factory(cache_type)

    storage_cache = StorageCache(
        redis_host, redis_port, redis_password, redis_db, key_format)

    expiration_timeout = 3600 * 24 * 7 * 2  # 2 weeks

    if options.sub_command == 'set-hash':
        secondary_keys = options.secondary_keys
        meta_data = json.dumps(options.meta_data)
        result = storage_cache.set_cache_hash(
            primary_keys, secondary_keys, meta_data, expiration_timeout)
    elif options.sub_command == 'get-hash':
        secondary_keys = options.secondary_keys
        result_str = storage_cache.get_cache_hash(
            primary_keys, secondary_keys)
        result = json.loads(result_str)
    elif options.sub_command == 'set-str':
        meta_data = json.dumps(options.meta_data)
        result = storage_cache.set_cache_str(
            primary_keys, meta_data, expiration_timeout)
    elif options.sub_command == 'get-str':
        result_str = storage_cache.get_cache_str(primary_keys)
        result = json.loads(result_str)

    LOGGER.info(result)
    LOGGER.info(type(result))


if __name__ == '__main__':
    logging.basicConfig(format=DEFAULT_LOG_FORMAT_STRING,
                        datefmt="%Y-%m-%d %H:%M:%S", level=logging.INFO)
    _main()


cloud_storage_cache/cache/redis_cache.py
"""
:author: Henley Kuang
:since: 06/13/2019
"""
import logging

from retry_redis import Redis

from cloud_storage_cache.util.config import (
    DEFAULT_LOG_FORMAT_STRING,
    DELIMITER,
)

LOGGER = logging.getLogger(__name__)


class CloudStorageRedisCache(object):

    redis_db = None
    key_format = None

    def __init__(self, redis_host, redis_port, redis_password, redis_db, key_format):
        self.key_format = key_format
        self.redis_client = Redis(host=redis_host,
                                  port=redis_port,
                                  password=redis_password,
                                  db=redis_db,
                                  decode_responses=True)

    def get_key_name(self, primary_keys):
        primary_keys_str = DELIMITER.join(primary_keys)
        key_name = "%s%s%s" % (self.key_format, DELIMITER, primary_keys_str)
        return key_name

    def set_cache_hash(self, primary_keys, secondary_keys, meta_data, expiration_timeout):
        key_name = self.get_key_name(primary_keys)
        field_key = DELIMITER.join(secondary_keys)
        # Hash set and set expiration pipeline
        rp = self.redis_client.pipeline()
        rp.hset(key_name, field_key, meta_data)
        rp.expire(key_name, expiration_timeout)
        rp.execute()
        return ('hset', key_name, field_key, meta_data)

    def get_cache_hash(self, primary_keys, secondary_keys):
        key_name = self.get_key_name(primary_keys)
        field_key = DELIMITER.join(secondary_keys)
        meta_data = self.redis_client.hget(key_name, field_key)
        return meta_data

    def set_cache_str(self, primary_keys, meta_data, expiration_timeout):
        key_name = self.get_key_name(primary_keys)
        self.redis_client.setex(key_name, expiration_timeout, meta_data)
        return ('set', key_name, meta_data)

    def get_cache_str(self, primary_keys):
        key_name = self.get_key_name(primary_keys)
        meta_data = self.redis_client.get(key_name)
        return meta_data


playground/cloud_storage_cache_playground.py
import argparse
import logging
import json

from cloud_storage_cache import CloudStorageCacheFactory
from cloud_storage_cache.util.config import DEFAULT_LOG_FORMAT_STRING

LOGGER = logging.getLogger(__name__)


def _parse_args():
    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawTextHelpFormatter)

    base_parser = argparse.ArgumentParser(add_help=False)
    base_parser.add_argument('--cache-type',  required=True)
    base_parser.add_argument('--redis-host', required=True)
    base_parser.add_argument('--redis-port', required=True)
    base_parser.add_argument('--redis-password')
    base_parser.add_argument('--redis-db', required=True)
    base_parser.add_argument('--key-format', required=True, help='key prefix')
    base_parser.add_argument('--primary-keys', nargs='+',
                             required=True, help='list of strings')

    hash_parser = argparse.ArgumentParser(add_help=False)
    hash_parser.add_argument('--secondary-keys', nargs='+',
                             required=True, help='list of strings')

    set_parser = argparse.ArgumentParser(add_help=False)
    set_parser.add_argument('--meta-data', nargs='+',
                            required=True, help='list of strings')

    command_subparser = parser.add_subparsers(dest='sub_command')
    command_subparser.required = True

    command_subparser.add_parser(
        'set-hash', parents=[base_parser, hash_parser, set_parser])
    command_subparser.add_parser(
        'get-hash', parents=[base_parser, hash_parser])
    command_subparser.add_parser('set-str', parents=[base_parser, set_parser])
    command_subparser.add_parser('get-str', parents=[base_parser])

    return parser.parse_args()


def _main():
    options = _parse_args()
    cache_type = options.cache_type
    redis_host = options.redis_host
    redis_port = options.redis_port
    redis_password = options.redis_password
    redis_db = options.redis_db
    key_format = options.key_format
    primary_keys = options.primary_keys

    CloudStorageCache = CloudStorageCacheFactory.factory(cache_type)

    cloud_storage_cache = CloudStorageCache(
        redis_host, redis_port, redis_password, redis_db, key_format)

    expiration_timeout = 3600 * 24 * 7 * 2  # 2 weeks

    if options.sub_command == 'set-hash':
        secondary_keys = options.secondary_keys
        meta_data = json.dumps(options.meta_data)
        result = cloud_storage_cache.set_cache_hash(
            primary_keys, secondary_keys, meta_data, expiration_timeout)
    elif options.sub_command == 'get-hash':
        secondary_keys = options.secondary_keys
        result_str = cloud_storage_cache.get_cache_hash(
            primary_keys, secondary_keys)
        result = json.loads(result_str)
    elif options.sub_command == 'set-str':
        meta_data = json.dumps(options.meta_data)
        result = cloud_storage_cache.set_cache_str(
            primary_keys, meta_data, expiration_timeout)
    elif options.sub_command == 'get-str':
        result_str = cloud_storage_cache.get_cache_str(primary_keys)
        result = json.loads(result_str)

    LOGGER.info(result)


if __name__ == '__main__':
    logging.basicConfig(format=DEFAULT_LOG_FORMAT_STRING,
                        datefmt="%Y-%m-%d %H:%M:%S", level=logging.INFO)
    _main()
